lib/jwt/jwa.rb

```diff
@@ -8,53 +8,37 @@
   raise if defined?(RbNaCl)
 end
 
+require_relative 'jwa/signing_algorithm'
 require_relative 'jwa/hmac'
 require_relative 'jwa/eddsa'
 require_relative 'jwa/ecdsa'
 require_relative 'jwa/rsa'
 require_relative 'jwa/ps'
 require_relative 'jwa/none'
-require_relative 'jwa/unsupported'
-require_relative 'jwa/wrapper'
+
+if JWT.rbnacl_6_or_greater?
+  require_relative 'jwa/hmac_rbnacl'
+elsif JWT.rbnacl?
+  require_relative 'jwa/hmac_rbnacl_fixed'
+end
 
 module JWT
   module JWA
-    ALGOS = [Hmac, Ecdsa, Rsa, Eddsa, Ps, None, Unsupported].tap do |l|
-      if ::JWT.rbnacl_6_or_greater?
-        require_relative 'jwa/hmac_rbnacl'
-        l << Algos::HmacRbNaCl
-      elsif ::JWT.rbnacl?
-        require_relative 'jwa/hmac_rbnacl_fixed'
-        l << Algos::HmacRbNaClFixed
-      end
-    end.freeze
-
     class << self
-      def find(algorithm)
-        indexed[algorithm&.downcase]
-      end
+      def resolve(algorithm)
+        return find(algorithm) if algorithm.is_a?(String) || algorithm.is_a?(Symbol)
 
-      def create(algorithm)
-        return algorithm if JWA.implementation?(algorithm)
-
-        Wrapper.new(*find(algorithm))
-      end
-
-      def implementation?(algorithm)
-        (algorithm.respond_to?(:valid_alg?) && algorithm.respond_to?(:verify)) ||
-          (algorithm.respond_to?(:alg) && algorithm.respond_to?(:sign))
+        wrap_algorithm(algorithm)
+        algorithm
       end
 
       private
 
-      def indexed
-        @indexed ||= begin
-          fallback = [nil, Unsupported]
-          ALGOS.each_with_object(Hash.new(fallback)) do |cls, hash|
-            cls.const_get(:SUPPORTED).each do |alg|
-              hash[alg.downcase] = [alg, cls]
-            end
-          end
+      def wrap_algorithm(algorithm)
+        return if algorithm.respond_to?(:header)
+
+        def algorithm.header(*)
```

Comment:
I wonder if it would be more clear if this would be in a separate module and extended to the given algorithm? We could maybe even use `JWT::JWA::SigningAlgorithm` directly, combined with a warning saying that algorithms needs to include this module to work in the future.